oslo/transformers/tasks/data_bart_pretraining.py

```diff
@@ -102,122 +142,149 @@ def __init__(
+        """
+        results = input_ids.copy()
+
+        # find end locations of sentences
+        # When creating an example, when a sentence comes in, sentence separation is performed, and the PAD token is inserted between the sentences.
+        end_sentence_mask = input_ids == self.tokenizer.pad_token_id
+        sentence_ends = np.argwhere(end_sentence_mask)
+        sentence_ends[:, 1] += 1
+        example_has_multiple_sentences, num_sentences = np.unique(sentence_ends[:, 0], return_counts=True)
+        
+        # E.X. num_sentences_map[3] = 10 indicates that there are 10 separate sentences in batch_index 3.
+        num_sentences_map = {sent_idx: count for sent_idx, count in zip(example_has_multiple_sentences, num_sentences)}
+        # Num_to_permute mixes only the sentence by the permit_ratio of the total when there are actually 10 sentences to be separated.
+        num_to_permute = np.ceil(num_sentences * self.permute_sentence_ratio).astype(int)
+        num_to_permute_map = {
+            sent_idx: count for sent_idx, count in zip(example_has_multiple_sentences, num_to_permute)
+        }
 
-        batch["decoder_input_ids"] = shifted_input_ids
-        batch["decoder_attention_mask"] = torch.where(
-            shifted_input_ids == self.pad_token_id,
-            0,
-            torch.ones_like(shifted_input_ids),
-        )
-        return batch
+        sentence_ends = np.split(sentence_ends[:, 1], np.unique(sentence_ends[:, 0], return_index=True)[1][1:])
+        sentence_ends_map = {sent_idx: count for sent_idx, count in zip(example_has_multiple_sentences, sentence_ends)}
+
+        for i in range(input_ids.shape[0]):
+            if i not in example_has_multiple_sentences:
+                continue
+            # Create a random number of sentences to be separated (num_sentences_map) and extract by the permit_ratio (num_to_permute_map).
+            # Substitutions is a random selection of sentences to be selected, and ordering[substitutions] is a random mixing of selected sentences.
+            substitutions = np.random.permutation(num_sentences_map[i])[: num_to_permute_map[i]]
+            ordering = np.arange(0, num_sentences_map[i])
+            # In orderring, mix the sentences by replacing the index corresponding to substitions with a random mixture of num_to_permute_map.
+            ordering[substitutions] = substitutions[np.random.permutation(num_to_permute_map[i])]
+
+            # write shuffled sentences into results
+            # Need to solve the problem of having to repeat the whole sentence, later.
+            index = 0
+            for j in ordering:
+                sentence = input_ids[i, (sentence_ends_map[i][j - 1] if j > 0 else 0) : sentence_ends_map[i][j]]
+                results[i, index : index + sentence.shape[0]] = sentence
+                index += sentence.shape[0]
+        return results  
+
+    def text_infilling(self, input_ids, labels, do_permute):
```

Comment:
text_infilling function - 

We randomly set tokens in sentences and mask them.
Determined how many tokens you want to extend according to the Poisson distribution.

I don't make special_token or pad_token masking.